File: adk/python/liate.py
"""
Liate ADK for Python
Sovereign AI Agent Client & Builder (LAPI/v1)
5-Pillars: L-I-A-T-E Architecture
"""

import os
import json
import urllib.request
import urllib.error
from typing import Dict, List, Any, Optional, Union

class LiateModel:
    def __init__(self, model: str = "sarvam/sarvam-105b", fallback: Optional[str] = None, temperature: Optional[float] = None):
        self.model = model
        self.fallback = fallback
        self.temperature = temperature

    def to_json(self) -> str:
        return self.model

class LiateIntegration:
    def __init__(self, memory: Optional[str] = None, session: Optional[str] = None, database: Optional[str] = None, webhook: Optional[str] = None, **kwargs):
        self.memory = memory or session
        self.session = session or memory
        self.database = database
        self.webhook = webhook
        self.extra = kwargs

    def to_json(self) -> Dict[str, Any]:
        data = {}
        if self.memory:
            data["memory"] = self.memory
        if self.session:
            data["session"] = self.session
        if self.database:
            data["database"] = self.database
        if self.webhook:
            data["webhook"] = self.webhook
        data.update(self.extra)
        return data

class LiateTools:
    def __init__(self, tools: Optional[List[Any]] = None):
        self.tools = tools or []

    def add(self, tool: Any) -> "LiateTools":
        self.tools.append(tool)
        return self

    def to_json(self) -> List[str]:
        return [t if isinstance(t, str) else getattr(t, "name", str(t)) for t in self.tools]

class LiateEnv:
    def __init__(self, MAX_TURNS: int = 5, REASONING_EFFORT: str = "medium", TIMEOUT_MS: Optional[int] = None, **kwargs):
        self.MAX_TURNS = MAX_TURNS
        self.REASONING_EFFORT = REASONING_EFFORT
        self.TIMEOUT_MS = TIMEOUT_MS
        self.extra = kwargs

    def to_json(self) -> Dict[str, str]:
        data = {
            "MAX_TURNS": str(self.MAX_TURNS),
            "REASONING_EFFORT": self.REASONING_EFFORT
        }
        if self.TIMEOUT_MS is not None:
            data["TIMEOUT_MS"] = str(self.TIMEOUT_MS)
        for k, v in self.extra.items():
            data[k] = str(v)
        return data

class AgentHandle:
    def __init__(self, agent_id: str, client: "LiateClient"):
        self.agent_id = agent_id
        self.client = client

    def run(self, prompt: str, session: Optional[str] = None, stream: bool = False) -> str:
        url = f"{self.client.base_url.rstrip('/')}/lapi/v1/{self.agent_id}/run"
        payload = json.dumps({
            "prompt": prompt,
            "session": session,
            "stream": stream
        }).encode("utf-8")

        headers = {"Content-Type": "application/json"}
        if self.client.api_key:
            headers["Authorization"] = f"Bearer {self.client.api_key}"

        req = urllib.request.Request(url, data=payload, headers=headers, method="POST")
        try:
            with urllib.request.urlopen(req) as resp:
                data = json.loads(resp.read().decode("utf-8"))
                return data.get("response") or data.get("output") or data.get("result") or str(data)
        except urllib.error.HTTPError as e:
            err = e.read().decode("utf-8")
            raise RuntimeError(f"[Liate ADK Error {e.code}]: {err}")

    def get(self) -> Dict[str, Any]:
        url = f"{self.client.base_url.rstrip('/')}/lapi/v1/{self.agent_id}"
        req = urllib.request.Request(url, method="GET")
        with urllib.request.urlopen(req) as resp:
            return json.loads(resp.read().decode("utf-8"))

class LiateClient:
    def __init__(self, base_url: Optional[str] = None, api_key: Optional[str] = None):
        self.base_url = base_url or os.getenv("LIATE_BASE_URL", os.getenv("LIATE_ENDPOINT", "http://localhost:7071"))
        self.api_key = api_key or os.getenv("LIATE_API_KEY")

    def agent(self, agent_id: str) -> AgentHandle:
        return AgentHandle(agent_id, self)

    def list_agents(self) -> List[Dict[str, Any]]:
        url = f"{self.base_url.rstrip('/')}/lapi/v1/agents"
        req = urllib.request.Request(url, method="GET")
        try:
            with urllib.request.urlopen(req) as resp:
                return json.loads(resp.read().decode("utf-8"))
        except Exception:
            return []

class LiateAgent:
    def __init__(
        self,
        L: Union[str, LiateModel] = "sarvam/sarvam-105b",
        I: Optional[Union[Dict[str, Any], LiateIntegration]] = None,
        A: Optional[Dict[str, Any]] = None,
        T: Optional[Union[List[Any], LiateTools]] = None,
        E: Optional[Union[Dict[str, Any], LiateEnv]] = None,
        endpoint: Optional[str] = None,
        api_key: Optional[str] = None,
    ):
        self.L = L if isinstance(L, LiateModel) else LiateModel(L)
        self.I = I if isinstance(I, LiateIntegration) else LiateIntegration(**(I or {}))
        self.A = A or {}
        self.T = T if isinstance(T, LiateTools) else LiateTools(T or [])
        self.E = E if isinstance(E, LiateEnv) else LiateEnv(**(E or {}))
        self.endpoint = endpoint or os.getenv("LIATE_ENDPOINT", "http://localhost:7071")
        self.api_key = api_key or os.getenv("LIATE_API_KEY")

    def run(self, prompt: str, session: Optional[str] = None) -> str:
        agent_name = self.A.get("name", "default")
        url = f"{self.endpoint.rstrip('/')}/lapi/v1/{agent_name}/run"
        spec = {
            "L": self.L.to_json(),
            "I": self.I.to_json(),
            "A": self.A,
            "T": self.T.to_json(),
            "E": self.E.to_json()
        }
        payload = json.dumps({
            "spec": spec,
            "prompt": prompt,
            "session": session or self.I.session
        }).encode("utf-8")

        headers = {"Content-Type": "application/json"}
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"

        req = urllib.request.Request(url, data=payload, headers=headers, method="POST")
        try:
            with urllib.request.urlopen(req) as resp:
                data = json.loads(resp.read().decode("utf-8"))
                return data.get("response") or data.get("output") or data.get("result") or str(data)
        except urllib.error.HTTPError as e:
            err = e.read().decode("utf-8")
            raise RuntimeError(f"[Liate ADK Error {e.code}]: {err}")

def liate(L: Union[str, LiateModel], I: Optional[Dict[str, Any]] = None, **kwargs) -> LiateAgent:
    return LiateAgent(L=L, I=I, **kwargs)

Agent = LiateAgent

File: adk/go/liate.go
package liate

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
)

type ToolServer struct {
	Command string   `json:"command,omitempty"`
	Args    []string `json:"args,omitempty"`
	Url     string   `json:"url,omitempty"`
	Tools   []string `json:"tools,omitempty"`
}

type Action struct {
	Name   string `json:"name"`
	Intent string `json:"intent,omitempty"`
	Skills string `json:"skills,omitempty"`
}

type Identity struct {
	Memory string `json:"memory,omitempty"`
	Name   string `json:"name,omitempty"`
}

type Pillars struct {
	L string                 `json:"L"`
	I interface{}            `json:"I,omitempty"`
	A interface{}            `json:"A,omitempty"`
	T map[string]ToolServer  `json:"T,omitempty"`
	E map[string]string      `json:"E,omitempty"`
}

type Agent struct {
	Spec     Pillars
	Endpoint string
	ApiKey   string
}

func NewAgent(spec Pillars, customEndpoint ...string) *Agent {
	ep := "http://localhost:7071"
	if len(customEndpoint) > 0 && customEndpoint[0] != "" {
		ep = customEndpoint[0]
	} else if envEp := os.Getenv("LIATE_ENDPOINT"); envEp != "" {
		ep = envEp
	} else if envEp := os.Getenv("LIATE_BASE_URL"); envEp != "" {
		ep = envEp
	}
	apiKey := os.Getenv("LIATE_API_KEY")
	return &Agent{Spec: spec, Endpoint: ep, ApiKey: apiKey}
}

func Om(spec Pillars, customEndpoint ...string) *Agent {
	return NewAgent(spec, customEndpoint...)
}

func (a *Agent) WithApiKey(key string) *Agent {
	a.ApiKey = key
	return a
}

func (a *Agent) Run(prompt string) (string, error) {
	payload, err := json.Marshal(map[string]interface{}{
		"spec":   a.Spec,
		"prompt": prompt,
	})
	if err != nil {
		return "", err
	}

	agentName := "default"
	if act, ok := a.Spec.A.(Action); ok && act.Name != "" {
		agentName = act.Name
	} else if actMap, ok := a.Spec.A.(map[string]interface{}); ok {
		if n, ok := actMap["name"].(string); ok && n != "" {
			agentName = n
		}
	}

	url := fmt.Sprintf("%s/lapi/v1/%s/run", strings.TrimRight(a.Endpoint, "/"), agentName)
	req, err := http.NewRequest("POST", url, bytes.NewBuffer(payload))
	if err != nil {
		return "", err
	}

	req.Header.Set("Content-Type", "application/json")
	if a.ApiKey != "" {
		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", a.ApiKey))
	}

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		// Fallback to /api/agent/run if /lapi/v1 was not reachable
		fallbackUrl := fmt.Sprintf("%s/api/agent/run", strings.TrimRight(a.Endpoint, "/"))
		reqFallback, fallbackErr := http.NewRequest("POST", fallbackUrl, bytes.NewBuffer(payload))
		if fallbackErr == nil {
			reqFallback.Header.Set("Content-Type", "application/json")
			if a.ApiKey != "" {
				reqFallback.Header.Set("Authorization", fmt.Sprintf("Bearer %s", a.ApiKey))
			}
			resp, err = client.Do(reqFallback)
		}
		if err != nil {
			return "", err
		}
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	if resp.StatusCode >= 400 {
		return "", fmt.Errorf("[Liate ADK Error %d]: %s", resp.StatusCode, string(body))
	}

	var data map[string]interface{}
	if err := json.Unmarshal(body, &data); err == nil {
		if out, ok := data["response"].(string); ok {
			return out, nil
		}
		if out, ok := data["output"].(string); ok {
			return out, nil
		}
		if out, ok := data["result"].(string); ok {
			return out, nil
		}
	}
	return string(body), nil
}
